src/fairseq2/recipes/evaluator.py

```diff
@@ -199,3 +203,181 @@ def _set_throughput(self, metric_values: Dict[str, Any]) -> None:
             metric_values["elements_per_second"] = 0.0
         else:
             metric_values["elements_per_second"] = num_elements / self._elapsed_time
+
+
+@final
+class HFEvaluator(Evaluator, Generic[BatchT]):
+    """Evaluate a machine learning model with HuggingFace's evaluate.Metric library"""
+
+    _model: Model
+    _root_gang: Gang
+    _dp_gang: Gang
+    _tp_gang: Gang
+    _data_reader: DataReader[BatchT]
+    _metric_recorders: List[MetricRecorder]
+    _wall_watch: Stopwatch
+    _elapsed_time: float
+    _run: bool
+
+    def __init__(
+        self,
+        model: Model,
+        metrics: List[str],
+        gang: Gang,
+        data_reader: DataReader[BatchT],
+        wall_watch: Stopwatch,
+        postprocess: Optional[Callable[[Sequence], Any]] = None,
+        dp_gang: Optional[Gang] = None,
+        tp_gang: Optional[Gang] = None,
+        tb_dir: Optional[Path] = None,
+    ) -> None:
+        """
+        :param model:
+            The fairseq2 machine learning model to be evaluate
+        :param metrics:
+            The list of metric names implemented in HuggingFace.evaluate
+        :param gang:
+            The gang to use for distributed evaluation.
+        :param data_reader:
+            The data reader of the eval split.
+        :param wall_watch:
+            The stopwatch to track process wall-time.
+        :param tb_dir:
+            The TensorBoard log directory to dump metrics.
+        """
+        try:
+            evaluate = importlib.import_module("evaluate")
+        except ImportError as exc:
+            raise ImportError(
+                "HFMetric requires the library `evaluate`, for instance via `pip install evaluate`"
+            ) from exc
+
+        self._model = model
+
+        self._root_gang = gang
+
+        if dp_gang is not None and tp_gang is not None:
+            self._dp_gang = dp_gang
+            self._tp_gang = tp_gang
+        elif dp_gang is None and tp_gang is None:
+            self._dp_gang = gang
+            self._tp_gang = FakeGang(device=gang.device)
+        else:
+            raise ValueError("`dp_gang` and `tp_gang` must be both specified.")
+
+        self._data_reader = data_reader
+
+        #########################################################
+        # FIXME: Implement the loading of evaluate metrics here,
+        # the remove the placeholder 'self,_metrics = None` below
+        #
+        #########################################################
+        self._metrics = [evaluate.load(metric) for metric in metrics]
```

Comment:
we can use "combine()" in HF eval https://huggingface.co/docs/evaluate/en/a_quick_tour#combining-several-evaluations

This will avoid iterating through the dataset multiple times for each metrics